### PR TITLE
[#18847][fix] Restore worker CPU affinity at shutdown

### LINKED ISSUE #18847: [Bug] Worker CPU affinity is applied process-wide in shared-process deployments and never restored
### Summary

`configure_cpu_affinity()` applies a *process-wide* CPU affinity policy, but in several supported configurations the worker does not own the process it runs in. In those deployments the mask is applied to threads that do not belong to the worker, and it is never restored when the worker shuts down.

This is a pre-existing condition, not a regression: before #18701 the code called `psutil.Process().cpu_affinity(cpus)`, which on Linux is `sched_setaffinity(pid, ...)` and rebinds the **main thread** of whatever process it runs in — in the shared cases below, that is the caller's own main thread. Because Linux threads inherit the creating thread's mask, everything the main thread spawned afterwards was pinned too. #18701 widened the blast radius from "the main thread and its later descendants" to "every TID present at the configuration point", which is what makes it worth tracking explicitly.

### Where the worker shares a process

All four reach `configure_cpu_affinity()` in a process that also hosts non-worker code:

1. **Single-process TP1 worker.** `executor.py` takes the `use_worker=True` branch when `gather_generation_logits=True` or `TLLM_WORKER_USE_SINGLE_PROCESS=1`, constructing `GenerationExecutorWorker` inline → `worker.py` `setup_engine()` → `base_worker.py:153`. Co-resident: the user's interpreter and its threads. Under `trtllm-serve` this is the HTTP server process itself — uvicorn/FastAPI, its uvloop event loop and the `OpenAIServer` thread pools all live there.
2. **Externally supplied `MpiCommSession`, rank 0.** `mpi_session.py` submits only `n_workers - 1` tasks to the MPI pool; rank 0's `worker_main` runs on the session's own `ThreadPoolExecutor` inside the submitting process. Co-resident: the user's rank-0 script and its event loop, the proxy's ZMQ dispatch/result threads, the second pool thread.
3. **External-launch VisualGen, rank 0.** `_torch/visual_gen/executor.py` starts the `_serve_forever` coordinator thread *before* the worker thread, then runs `configure_cpu_affinity()` on the worker thread at `executor.py:687`. Co-resident: the coordinator's own asyncio loop and ZMQ sockets, and under `trtllm-serve` the HTTP server on the main thread.
4. **Leader process of a multi-node launch.** With `TLLM_SPAWN_PROXY_PROCESS=1` the leader runs `RemoteMpiCommSessionServer.serve()` *and* rank 0's worker on the session thread pool. Most threads there are genuinely worker threads, but the ZMQ session server is not.

For contrast, these are dedicated and unaffected: Ray actors (`ray/gpu_worker.py:370`), ordinary MPI pool workers, and `mgmn_worker_node` ranks > 0.

### Problems

- **Ownership.** There is no signal distinguishing "this process is mine" from "I am a guest in the caller's process", so the policy cannot be scoped correctly today.
- **No restoration.** Nothing restores the previous mask at shutdown — `grep -rn "cpu_affinity\|sched_setaffinity" tensorrt_llm/` finds only the three `configure_cpu_affinity()` call sites. A user who constructs an `LLM`, uses it and shuts it down is left with a permanently narrowed process.
- **Default-on widening.** The `TLLM_NUMA_AWARE_WORKER_AFFINITY`-unset branch *removes* an externally set constraint. Under `srun --cpu-bind`, this now widens the whole hosting process to all CPUs rather than just its main thread.

### Suggested direction

`_torch/visual_gen/executor.py` already plumbs an `in_client_process` flag through `run_diffusion_worker`, so VisualGen is the cheapest first landing point — but the fix should cover all four modes rather than special-casing one.

Some care is needed: gating the sweep on "dedicated worker process" would silently disable NUMA pinning in the multi-node leader, where rank 0's full engine lives. Rank 0's MPI/UCX/NCCL progress threads would stay unpinned while every other rank is pinned, reproducing exactly the rank-0 straggler imbalance the feature exists to remove. "Restrict to worker-owned threads" is also not directly implementable: the threads the feature targets are unregistered C-level threads with no ownership metadata.

### Context

Raised in review of #18701, which documents the current behaviour in `docs/source/deployment-guide/configuring-cpu-affinity.md` but deliberately does not change it.

### PR DESCRIPTION
**Pull request by DOCCA0**
## Description

Closes #18847.

Worker CPU affinity can affect application threads when the worker shares the caller's process, and the old mask stays in place after shutdown.

Save the original per-thread masks before applying NUMA affinity and restore them when the last worker shuts down. The cleanup also runs when setup or shutdown fails. If another component changes a mask while the worker is running, leave that mask alone.

The worker, Ray and VisualGen paths are covered.

## Test Coverage

- CPU affinity lifecycle tests, including real Linux affinity syscalls (`12 passed`)
- Setup/shutdown failure and VisualGen worker tests
- Manual A/B test on four Quadro RTX 6000 nodes
- Pre-commit checks

## PR Checklist

- [x] Reviewed the items in the PR template.

## Dev Engineer Review

- Added CPU-affinity lease APIs that restore masks on final release.
- Worker, Ray, and VisualGen paths release leases during normal and failure cleanup.
- `set_postproc_queues()` now supports `coexist_with_result_queue`; RPC workers use it for shared queue flows.
- Verify affinity-operation failures and compatibility with existing postprocessing callers.

## QA Engineer Review

- Added unit tests for lease restoration, concurrent leases, external mask changes, idempotent release, and affinity failures.
- Added worker setup and shutdown failure-path tests.
- Updated VisualGen tests for lease handling and process-group initialization failure.
- No test-list files changed.
- Coverage verdict: **sufficient**.

## Per-File QA Perspective

- `tensorrt_llm/_torch/visual_gen/executor.py`: Verify lease cleanup on worker errors and external-launch rank 0 operation.
- `tensorrt_llm/executor/base_worker.py`: Verify lease release on setup and shutdown failures. Verify classic and RPC postprocessing queue behavior.
- `tensorrt_llm/executor/ray/gpu_worker.py`: Verify cleanup when Ray shutdown or postprocessing fails.
- `tensorrt_llm/executor/worker.py`: Verify release when teardown raises.
- `tensorrt_llm/llmapi/utils.py`: Verify concurrent leases, fork behavior, external mask changes, and affinity-operation failures.
- `tests/unittest/_torch/visual_gen/multi_gpu/test_visual_gen_multinode.py`: Covers device assignment, lease handling, and process-group initialization failure. No test-list entry is required.
- `tests/unittest/executor/test_base_worker.py`: Covers lease release after setup or shutdown errors. No test-list entry is required.
- `tests/unittest/llmapi/test_cpu_affinity_lifecycle.py`: Covers lease restoration and failure behavior, including Linux thread-affinity restoration. No test-list entry is required.

### CODE REVIEW

File: tensorrt_llm/_torch/visual_gen/executor.py [hunk 1]
```diff
@@ -835,6 +836,8 @@ def run_diffusion_worker(
         logger.error(f"Worker failed: {e}")
         traceback.print_exc()
         raise
+    finally:
+        release_cpu_affinity(cpu_affinity_lease)
```
[the remark this answers is not in the document]
Comment (DOCCA0) [reply tensorrt_llm/_torch/visual_gen/executor.py]: Added a failure-path case where init_process_group raises after the lease is acquired. It checks that release_cpu_affinity(7) still runs.

File: tensorrt_llm/executor/base_worker.py [hunk 2]
```diff
@@ -1025,8 +1039,13 @@ def shutdown(self):
...
                 can_shutdown = self.engine.can_enqueue_requests
             if can_shutdown():
-                self.engine.shutdown()
+                try:
+                    self.engine.shutdown()
+                finally:
+                    self._release_cpu_affinity()
                 self.engine = None
+        else:
+            self._release_cpu_affinity()
```
[the remark this answers is not in the document]
Comment (DOCCA0) [reply tensorrt_llm/executor/base_worker.py]: Fixed. BaseWorker.shutdown now releases the lease from a finally block, including false and exceptional can_shutdown() paths. Both cases have regression coverage.

File: tests/unittest/executor/test_base_worker.py [hunk 3]
```diff
@@ -109,6 +169,7 @@ def shutdown(self):
         if self.engine is not None:
             self.engine.shutdown()
             self.engine = None
+        self._release_cpu_affinity()
```
[the remark this answers is not in the document]
Comment (DOCCA0) [reply tests/unittest/executor/test_base_worker.py]: Fixed the test worker cleanup to use finally around engine.shutdown().

File: tests/unittest/llmapi/test_cpu_affinity.py [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+def test_last_affinity_lease_restores_existing_and_later_threads(affinity_leases):
+    current, bindings, all_cpus, numa_cpus = affinity_leases
+    lease = utils.acquire_cpu_affinity(0)
+    current[103] = numa_cpus
+
+    utils.release_cpu_affinity(lease)
+
+    assert current == {101: all_cpus, 102: frozenset({2, 3}), 103: all_cpus}
+    assert bindings == [(101, all_cpus), (102, frozenset({2, 3})), (103, all_cpus)]
+    assert utils._cpu_affinity_lease_state is None
```
[the remark this answers is not in the document]
Comment (DOCCA0) [reply tests/unittest/llmapi/test_cpu_affinity.py]: Added coverage for an unavailable initial snapshot and for the psutil restoration/preservation fallback when sched_setaffinity is unavailable.

File: tensorrt_llm/llmapi/utils.py [hunk 5]
```diff
@@ -710,6 +711,138 @@ def _reapply_current_thread_affinity_to_all_threads() -> tuple[int, int]:
...
+            except (AttributeError, OSError):
+                default_mask = frozenset(psutil.Process(pid).cpu_affinity())
+            state = _CpuAffinityLeaseState(pid, original_masks, default_mask)
+            before = _read_thread_affinities()
+            configure_cpu_affinity(device_id)
+            after = _read_thread_affinities()
+            state.owned_masks.update(mask for tid, mask in after.items()
+                                     if before.get(tid) != mask)
+            if not state.owned_masks:
+                return None
```
[the remark this answers is not in the document]
Comment (DOCCA0) [reply tensorrt_llm/llmapi/utils.py]: Fixed both cases. A failed post-configuration snapshot restores the saved TIDs before returning, and a failed final snapshot uses the same best-effort fallback before clearing the lease state. Added regression tests for both paths.

**Comment by DOCCA0** [comment]
Addressed the review feedback in db125e7e53:

- release affinity on every BaseWorker shutdown exit
- restore saved masks when affinity snapshots fail
- cover the psutil fallback
- add VisualGen and test-worker failure-path coverage

The focused affinity suite passes (11 tests), and pre-commit is clean. Ready for another look.

**Comment by DOCCA0** [comment]
@mikeiovine 
5 CodeRabbit thread resolved, 
thanks

File: tensorrt_llm/llmapi/utils.py [hunk 6]
```diff
@@ -710,6 +711,173 @@ def _reapply_current_thread_affinity_to_all_threads() -> tuple[int, int]:
...
+        if skipped:
+            logger.info(
+                f"Preserved externally changed CPU affinity on {skipped} threads."
+            )
+        logger.info(f"Restored pre-worker CPU affinity on {restored} threads.")
+        _cpu_affinity_lease_state = None
+
+
 def configure_cpu_affinity(device_id: int) -> None:
     """Probe and configure the CPU affinity of the calling process based on NUMA topology.
```
[the remark this answers is not in the document]
Comment (DOCCA0) [reply tensorrt_llm/llmapi/utils.py]: Good catch. Switched the lease lock to threading.RLock and added a regression test that acquires it recursively without blocking. The focused affinity suite now passes 12 tests, and pre-commit is clean.
